internal/controller/infrastructure/provisioner.go

```diff
@@ -0,0 +1,175 @@
+	"gopkg.in/yaml.v3"
+
+	// anonymous import is needed to load the os configurers
+	_ "github.com/k0sproject/k0sctl/configurer"
+	// anonymous import is needed to load the os configurers
+	_ "github.com/k0sproject/k0sctl/configurer/linux"
+	// anonymous import is needed to load the os configurers
+	_ "github.com/k0sproject/k0sctl/configurer/linux/enterpriselinux"
+)
+
+type Provisioner struct {
+	bootstrapData []byte
+	machine       *api.RemoteMachine
+	sshKey        []byte
+	log           logr.Logger
+}
+
+// Provision provisions a new machine
+// The provisioning process is as follows:
+// 1. Open SSH connection to the machine
+// 2. Execute the bootstrap script
+// 3. Check sentinel file at /run/cluster-api/bootstrap-success.complete
+// 4. success
+func (p *Provisioner) Provision(ctx context.Context) error {
+	// Parse the bootstrap data
+	cloudInit := &cloudinit.CloudInit{}
+	err := yaml.Unmarshal(p.bootstrapData, cloudInit)
+	if err != nil {
+		return fmt.Errorf("failed to parse bootstrap data: %w", err)
+	}
+
+	// Dump key to a temporary file
+	f, err := os.CreateTemp("", "k0smotron-ssh-key")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary key file: %w", err)
+	}
+	defer os.Remove(f.Name())
+
+	if _, err := f.Write(p.sshKey); err != nil {
+		return fmt.Errorf("failed to write key to temporary file: %w", err)
+	}
+	f.Close()
+
+	keyFile := f.Name()
+
+	connection := rig.Connection{
+		SSH: &rig.SSH{
+			Address: p.machine.Spec.Address,
+			Port:    p.machine.Spec.Port,
+			User:    p.machine.Spec.User,
+			KeyPath: &keyFile,
+		},
+	}
+
+	// Wrap the connection in k0sctl host to get OS resolving etc.
+	host := &k0sctl.Host{
+		Connection: connection,
+	}
+
+	if err := p.connect(host); err != nil {
+		return fmt.Errorf("failed to connect to host: %w", err)
+	}
+
+	defer host.Disconnect()
+
+	// Write files first
+	for _, file := range cloudInit.Files {
+		if err := p.uploadFile(host, file); err != nil {
+			return fmt.Errorf("failed to upload file: %w", err)
+		}
+	}
+
+	// Execute the bootstrap script commands
+	for _, cmd := range cloudInit.RunCmds {
+		output, err := host.ExecOutput(cmd)
+		if err != nil {
+			p.log.Error(err, "failed to run command", "output", output)
+			return fmt.Errorf("failed to run command: %w", err)
+		}
+	}
+
+	// Check for sentinel file
+	if !host.Configurer.FileExist(host, "/run/cluster-api/bootstrap-success.complete") {
+		return errors.New("bootstrap sentinel file not found")
+	}
+
+	return nil
+}
+
+const retries = uint(60)
+
+func (p *Provisioner) uploadFile(h *k0sctl.Host, file cloudinit.File) error {
+
+	// Ensure base dir exists for target
+	dir := filepath.Dir(file.Path)
+	if err := h.Configurer.MkDir(h, dir, exec.Sudo(h)); err != nil {
+		return fmt.Errorf("failed to create directory: %w", err)
+	}
+
+	if err := h.Configurer.WriteFile(h, file.Path, file.Content, file.Permissions); err != nil {
+		return fmt.Errorf("failed to write file: %w", err)
+	}
+
```

Comment:
If only using `Configurer.FileExist`,/`.Mkdir`/`.WriteFile`, you can get rid of the k0sctl dependency by using rig's fs.FS -stuff.

```go

// host := rig.Connection{...}
fsys := host.SudoFSys()

if _, err := fsys.Stat("/path/to/whatever"); err != nil {
  return fmt.Errorf("bootstrap sentinel file not found: %w", err)
}

// permissions is int
if err := fsys.MkDirAll(filepath.Dir(file.Path), 0700); err != nil {
  return fmt.Errorf("failed to create directory: %w", err)
}

dest, err := fsys.OpenFile(file.Path, rigfs.ModeCreate, file.Permissions)
if err != nil {
  return fmt.Errorf("failed to create file: %w", err)
}
defer dest.Close()

if _, err := io.WriteString(dest, file.Content); err != nil {
  return fmt.Errorf("failed to write file: %w", err)
}
```